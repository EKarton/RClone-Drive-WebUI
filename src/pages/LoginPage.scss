.login-page {
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100%;

  &__header {
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    padding: 1rem;
    text-align: right;
  }

  &__card {
    width: 23rem;
    padding: 1rem;
  }

  &__card-subheader {
    font: 1rem;

    &--red {
      color: red;
    }
  }

  &__card-contents {
    display: flex;
    row-gap: 1rem;
    flex-direction: column;
  }

  &__text-field {
    display: block;
  }
}
