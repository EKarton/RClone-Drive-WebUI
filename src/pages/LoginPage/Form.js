import LoginIcon from '@mui/icons-material/Login';
import LoadingButton from '@mui/lab/LoadingButton';
import Button from '@mui/material/Button';
import Card from '@mui/material/Card';
import CardActions from '@mui/material/CardActions';
import CardContent from '@mui/material/CardContent';
import CardHeader from '@mui/material/CardHeader';
import TextField from '@mui/material/TextField';
import cx from 'classnames';
import { useLoginHelpDialog } from 'contexts/LoginHelpDialog';
import './Form.scss';

export default function Form({ error, isLoading, onFormSubmit }) {
  const { openDialog } = useLoginHelpDialog();
  const hasError = error !== null;

  const cardSubHeader = (
    <span className={cx({ 'login-page-form__header--red': error })}>
      {hasError ? error?.message : 'Enter your RClone info to get started'}
    </span>
  );

  return (
    <Card variant="outlined" className="login-page-form">
      <form onSubmit={onFormSubmit}>
        <CardHeader title="Login" subheader={cardSubHeader} />
        <CardContent className="login-page-form__contents">
          <TextField
            required
            name="endpoint"
            label="RClone RC Endpoint"
            variant="outlined"
            placeholder="Ex: http://localhost:5572"
            className="login-page-form__text-field"
            error={hasError}
            inputProps={{ 'data-testid': 'endpoint-field' }}
          />
          <TextField
            required
            name="username"
            label="Username"
            variant="outlined"
            className="login-page-form__text-field"
            error={hasError}
            inputProps={{ 'data-testid': 'username-field' }}
          />
          <TextField
            required
            name="password"
            label="Password"
            variant="outlined"
            type="password"
            className="login-page-form__text-field"
            error={hasError}
            inputProps={{ 'data-testid': 'password-field' }}
          />
        </CardContent>
        <CardActions>
          <LoadingButton
            loading={isLoading}
            loadingPosition="start"
            startIcon={<LoginIcon />}
            variant="contained"
            data-testid="login-button"
            type="submit"
          >
            Login
          </LoadingButton>
          <Button data-testid="help-button" onClick={openDialog}>
            Help
          </Button>
        </CardActions>
      </form>
    </Card>
  );
}
