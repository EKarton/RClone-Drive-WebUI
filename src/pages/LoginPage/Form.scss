.login-page-form {
  width: 23rem;
  padding: 1rem;

  &__header {
    &--red {
      color: red;
    }
  }

  &__contents {
    display: flex;
    row-gap: 1rem;
    flex-direction: column;
  }

  &__text-field {
    display: block;
  }
}
