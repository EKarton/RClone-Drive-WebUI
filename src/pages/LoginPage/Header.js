import DarkModeToggleSwitch from 'components/DarkModeToggleSwitch';
import './Header.scss';

export default function Header() {
  return (
    <div className="login-page-header">
      <DarkModeToggleSwitch />
    </div>
  );
}
