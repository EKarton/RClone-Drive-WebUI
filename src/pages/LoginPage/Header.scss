.login-page-header {
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  padding: 1rem;
  text-align: right;
}
