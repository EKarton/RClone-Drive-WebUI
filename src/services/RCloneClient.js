import axios from 'axios';
import { getFullPath } from 'utils/filename-utils';

const ImageIncludeRules = [
  '*.png',
  '*.PNG',
  '*.jpg',
  '*.JPG',
  '*.jpeg',
  '*.JPEG',
  '*.bmp',
  '*.BMP',
  '*.gif',
  '*.GIF',
  '*.heic',
  '*.HEIC',
];

/**
 * A class used to handle all requests to RClone
 */
export default class RCloneClient {
  constructor(rCloneInfo) {
    this.axiosInstance = axios.create({
      responseType: 'json',
      baseURL: rCloneInfo.endpoint,
      headers: {
        'Content-Type': 'application/json',
      },
      auth: {
        username: rCloneInfo.username,
        password: rCloneInfo.password,
      },
    });
  }

  /**
   * Fetches and returns a list of remotes
   * @returns {Promise<string[]>} a list of remotes
   */
  async fetchRemotes(opts = {}) {
    const { data } = await this.axiosInstance.post('config/listremotes', undefined, opts);
    return data.remotes;
  }

  /**
   * Fetches a list of files and folders under a directory path in a remote
   * It returns a list of objects, where each object has this shape:
   *
   * {
   *   "Path": <the full file path of the folder / file >,
   *   "Name": <the name of the folder or file>,
   *   "Size": <the size of the folder or file, in bytes>,
   *   "MimeType": <the type (ex: "inode/directory")>,
   *   "ModTime": <the time this folder / file was updated, in ISO 8601 format>,
   *   "IsDir": <true if it is a directory; else false>,
   *   "ID": <the id of this object>
   * }
   *
   * @param {string} remote the remote
   * @param {string} directoryPath the directory path
   * @param {object} opts the set of axios options
   * @returns {Promise<object[]>} a list of files
   */
  async fetchFiles(remote, directoryPath, opts = {}) {
    const postData = {
      fs: `${remote}:`,
      remote: `${directoryPath}`,
      _config: {
        UseListR: true,
      },
    };
    const { data } = await this.axiosInstance.post('operations/list', postData, opts);

    return data.list;
  }

  /**
   * Fetches a list of folders under a directory path in a remote
   * It returns a list of objects, where each object has this shape:
   *
   * {
   *   "Path": <the full file path of the folder >,
   *   "Name": <the name of the folder or file>,
   *   "Size": <the size of the folder or file, in bytes>,
   *   "MimeType": <the type (ex: "inode/directory")>,
   *   "ModTime": <the time this folder / file was updated, in ISO 8601 format>,
   *   "IsDir": true,
   *   "ID": <the id of this object>
   * }
   *
   * @param {string} remote the remote
   * @param {string} directoryPath the directory path
   * @param {object} opts the set of axios options
   * @returns {Promise<object[]>} a list of files
   */
  async fetchSubFolders(remote, directoryPath, opts = {}) {
    const requestBody = {
      fs: `${remote}:`,
      remote: `${directoryPath}`,
      opt: {
        dirsOnly: true,
      },
    };
    const { data } = await this.axiosInstance.post('operations/list', requestBody, opts);

    return data.list;
  }

  /**
   * Fetches a list of pictures under a directory path in a remote
   * It returns a list of objects, where each object has this shape:
   *
   * {
   *   "Path": <the current path of the folder / file >,
   *   "Name": <the name of the folder or file>,
   *   "Size": <the size of the folder or file, in bytes>,
   *   "MimeType": <the type (ex: "image/png")>,
   *   "ModTime": <the time this folder / file was updated, in ISO 8601 format>,
   *   "IsDir": false,
   *   "ID": <the id of this object>
   * }
   *
   * @param {string} remote the remote
   * @param {string} directoryPath the directory path
   * @returns {Promise<object[]>} a list of files
   */
  async fetchPictures(remote, directoryPath, opts = {}) {
    const requestBody = {
      fs: `${remote}:`,
      remote: `${directoryPath}`,
      opt: {
        recurse: true,
        filesOnly: true,
      },
      _config: {
        UseListR: true,
      },
      _filter: {
        IncludeRule: ImageIncludeRules,
      },
    };

    const { data } = await this.axiosInstance.post('operations/list', requestBody, opts);

    return data.list;
  }

  /**
   * Returns the file contents of a file in a remote
   *
   * @param {string} remote the name of the remote
   * @param {string} directoryPath the folder path
   * @param {string} fileName the file name
   * @returns {Promise<string>} the blob contents of the file
   */
  async fetchFileContents(remote, directoryPath, fileName, opts = {}) {
    const { cancelToken } = opts;
    const remotePath = `${remote}:${directoryPath}`;
    const url = encodeURI(`[${remotePath}]/${fileName}`);

    const response = await this.axiosInstance.get(url, {
      responseType: 'blob',
      cancelToken,
    });

    return response;
  }

  /**
   * Returns the space info of a remote
   * The response returned is:
   * {
   *   "free": number,
   *   "other": number,
   *   "total": number,
   *   "trashed": number,
   *   "used": number
   *  }
   *
   * All numbers are in bytes
   *
   * @param {string} remote the remote
   * @returns {Promise<object>} the space info
   */
  async fetchRemoteSpaceInfo(remote, opts = {}) {
    const requestBody = {
      fs: `${remote}:`,
    };
    const { data } = await this.axiosInstance.post('operations/about', requestBody, opts);

    return data;
  }

  /**
   * Returns information about a full path in a remote
   * It returns information in this shape:
   * {
   *   "Path": string,
   *   "Name": string,
   *   "Size": integer,
   *   "MimeType": string,
   *   "ModTime": string,
   *   "IsDir": boolean,
   *   "ID": string
   * }
   *
   * If the full path doesn't exist in the remote, it will return null.
   * If the remote does not exist, it will throw an exception
   *
   * @param {string} remote the remote
   * @param {string} fullPath the file path
   * @param {string} opts options
   * @returns {Promise<object>} information about the file
   */
  async fetchFullPathInfo(remote, fullPath, opts = {}) {
    const requestBody = {
      fs: `${remote}:`,
      remote: fullPath,
    };

    const { data } = await this.axiosInstance.post('operations/stat', requestBody, opts);

    if (data.item === null) {
      throw new Error(`Cannot find full path ${fullPath}`);
    }

    return data.item;
  }

  /**
   * Returns the remote info
   * It will return an object with this shape:
   * {
   *    "scope": string,
   *    "type": string
   * }
   *
   * Depending on the type of remote, it may include more information.
   * For instance, with GDrives, it may return an object with this shape:
   * {
   *    "scope": string,
   *    "team_drive": string,
   *    "token": object,
   *    "type": "drive"
   * }
   *
   * @param {string} remote the remote
   * @returns {Promise<object>} the info of the remote
   */
  async fetchRemoteInfo(remote, opts = {}) {
    const requestBody = {
      name: remote,
    };
    const { data } = await this.axiosInstance.post('config/get', requestBody, opts);

    return data;
  }

  /**
   * Uploads a file to a folder
   *
   * @param {string} remote the remote
   * @param {string} dirPath the path to the folder
   * @param {File} data the file data
   * @param {Object?} opts the options
   */
  async uploadFiles(remote, dirPath, data, opts = {}) {
    const formData = new FormData();
    formData.append('file', data);

    const url = `operations/uploadfile?fs=${remote}:&remote=${dirPath}`;

    await this.axiosInstance.post(url, formData, opts);
  }

  /**
   * Deletes a directory from a remote
   * @param {string} remote string
   * @param {string} directoryPath the folder path
   * @param {string} baseName the name of the folder to delete
   */
  async deleteDirectory(remote, directoryPath, baseName, opts = {}) {
    const { isAsync = false } = opts;

    return this.axiosInstance.post('operations/purge', {
      fs: `${remote}:`,
      remote: getFullPath(directoryPath, baseName),
      _async: isAsync,
    });
  }

  async deleteFile(remote, dirPath, fileName, opts = {}) {
    const { isAsync = false } = opts;

    return this.axiosInstance.post('operations/deletefile', {
      fs: `${remote}:`,
      remote: getFullPath(dirPath, fileName),
      _async: isAsync,
    });
  }

  /**
   * Copies the contents in a source directory to a target directory
   * @param {object} source the source directory
   * @param {object} target the target directory
   * @param {boolean} createEmptySrcDirs true if it creates empty sub-directories in the source path; else false
   */
  async copyDirectoryContents(source, target, createEmptySrcDirs) {
    const srcPath = getFullPath(source.dirPath, source.fileName);
    const targetPath = getFullPath(target.dirPath, target.fileName);

    return this.axiosInstance.post('sync/copy', {
      srcFs: `${source.remote}:${srcPath}`,
      dstFs: `${target.remote}:${targetPath}`,
      createEmptySrcDirs,
    });
  }

  /**
   * Copies a file to a new path / remote
   * The source path must be of this shape:
   * {
   *    remote: string (the src remote),
   *    dirPath: string (the folder path of the src file)
   *    fileName: string (the src file name),
   * }
   *
   * The target path must be of this shape:
   * {
   *    remote: string (the new remote),
   *    dirPath: string (the new folder path of the file)
   *    fileName: string (the new file name)
   * }
   *
   * @param {object} source the source file, with the shape above
   * @param {object} target the target file, with the shape above
   */
  async copyFile(source, target) {
    return this.axiosInstance.post('operations/copyfile', {
      srcFs: `${source.remote}:`,
      srcRemote: getFullPath(source.dirPath, source.fileName),
      dstFs: `${target.remote}:`,
      dstRemote: getFullPath(target.dirPath, target.fileName),
    });
  }

  /**
   * Moves a file to a new path / remote
   * The source path must be of this shape:
   * {
   *    remote: string (the src remote),
   *    dirPath: string (the folder path of the src file)
   *    name: string (the src file name),
   * }
   *
   * The target path must be of this shape:
   * {
   *    remote: string (the new remote),
   *    dirPath: string (the new folder path of the file)
   *    name: string (the new file name)
   * }
   *
   * @param {object} source the source file, with the shape above
   * @param {object} target the target file, with the shape above
   */
  async moveFile(source, target, opts = {}) {
    const { isAsync = false } = opts;

    return this.axiosInstance.post('operations/movefile', {
      srcFs: `${source.remote}:`,
      srcRemote: getFullPath(source.dirPath, source.name),
      dstFs: `${target.remote}:`,
      dstRemote: getFullPath(target.dirPath, target.name),
      _async: isAsync,
    });
  }

  /**
   * Moves a directory from one path to another path
   * The source path must be of this shape:
   * {
   *    remote: string (the src remote),
   *    dirPath: string (the folder path of the src file)
   *    name: string (the name of the folder),
   * }
   *
   * The target path must be of this shape:
   * {
   *    remote: string (the new remote),
   *    dirPath: string (the new folder path of the folder)
   *    name: string (the new folder name)
   * }
   *
   * If the target's folder path does not exist, it will make one
   *
   * @param {object} source the source file, with the shape above
   * @param {object} target the target file, with the shape above
   */
  async move(source, target, opts = {}) {
    const { createEmptySrcDirs = false, deleteEmptySrcDirs = false } = opts;
    const { isAsync = false } = opts;
    const srcRemote = getFullPath(source.dirPath, source.name);
    const targetRemote = getFullPath(target.dirPath, target.name);

    return this.axiosInstance.post('sync/move', {
      srcFs: `${source.remote}:${srcRemote}`,
      dstFs: `${target.remote}:${targetRemote}`,
      createEmptySrcDirs,
      deleteEmptySrcDirs,
      _async: isAsync,
    });
  }

  /**
   * Makes a directory in a remote
   * @param {string} remote the remote
   * @param {string} fullPath the full path of the new folder
   */
  async mkdir(remote, fullPath) {
    return this.axiosInstance.post('operations/mkdir', {
      fs: `${remote}:`,
      remote: fullPath,
    });
  }

  /**
   * Empties the trash can
   * @param {string} remote the remote name
   */
  async emptyTrashCan(remote) {
    return this.axiosInstance.post('operations/cleanup', {
      fs: `${remote}:`,
    });
  }

  async getJobStatus(jobId) {
    return this.axiosInstance.post('job/status', {
      jobid: jobId,
    });
  }

  async stopJob(jobId) {
    return this.axiosInstance.post('job/stop', {
      jobid: jobId,
    });
  }
}
