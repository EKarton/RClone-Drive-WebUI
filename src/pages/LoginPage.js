import {
  Button,
  Card,
  CardActions,
  CardContent,
  CardHeader,
  TextField,
} from '@mui/material';
import cx from 'classnames';
import useRCloneInfo from 'hooks/useRCloneInfo';
import { useState } from 'react';
import { useHistory } from 'react-router';
import RCloneClient from 'utils/RCloneClient';
import './LoginPage.scss';

/**
 * The login page
 */
const LoginPage = () => {
  const { setRCloneInfo } = useRCloneInfo();
  const history = useHistory();

  const [endpoint, setEndpoint] = useState(null);
  const [username, setUsername] = useState(null);
  const [password, setPassword] = useState(null);

  const [error, setError] = useState(null);

  const handleTextEntered = (setFn) => (event) => {
    setFn(event.target.value);
  };

  const handleLogin = async () => {
    try {
      // To test if connection is correct, call RClone to fetch remotes
      await new RCloneClient(endpoint, username, password).fetchRemotes();

      setRCloneInfo({ endpoint, username, password });
      history.push('/files');
    } catch (error) {
      setError(error);
    }
  };

  const cardSubHeader = (
    <span className={cx({ 'login-page__card-subheader--red': error })}>
      {error ? error?.message : 'Enter your RClone info to get started'}
    </span>
  );

  const hasError = error === null;

  return (
    <div className="login-page">
      <Card variant="outlined" className="login-page__card">
        <CardHeader title="Login" subheader={cardSubHeader} />
        <CardContent className="login-page__card-contents">
          <TextField
            required
            label="RClone RC Endpoint"
            variant="outlined"
            placeholder="Ex: http://localhost:5572"
            className="login-page__text-field"
            onChange={handleTextEntered(setEndpoint)}
            error={hasError}
            inputProps={{ 'data-testid': 'endpoint-field' }}
          />
          <TextField
            required
            label="Username"
            variant="outlined"
            className="login-page__text-field"
            onChange={handleTextEntered(setUsername)}
            error={hasError}
            inputProps={{ 'data-testid': 'username-field' }}
          />
          <TextField
            required
            label="Password"
            variant="outlined"
            type="password"
            className="login-page__text-field"
            onChange={handleTextEntered(setPassword)}
            error={hasError}
            inputProps={{ 'data-testid': 'password-field' }}
          />
        </CardContent>
        <CardActions>
          <Button variant="contained" onClick={handleLogin} data-testid="login-button">
            Login
          </Button>
        </CardActions>
      </Card>
    </div>
  );
};

export default LoginPage;
