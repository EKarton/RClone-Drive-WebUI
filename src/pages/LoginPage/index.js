import Box from '@mui/material/Box';
import { useState } from 'react';
import { useNavigate, useSearchParams } from 'react-router-dom';
import { LoginHelpDialogProvider } from 'contexts/LoginHelpDialog';
import useRCloneInfo from 'hooks/rclone/useRCloneInfo';
import RCloneClient from 'services/RCloneClient';
import Form from './Form';
import Header from './Header';
import './index.scss';

export default function LoginPage() {
  const [searchParams] = useSearchParams();
  const navigate = useNavigate();
  const { setRCloneInfo } = useRCloneInfo();
  const [error, setError] = useState(null);
  const [isLoading, setIsLoading] = useState(false);

  const handleFormSubmit = async (e) => {
    e.preventDefault();

    const formProps = Object.fromEntries(new FormData(e.target));
    const { endpoint, username, password } = formProps;

    const redirectPath = searchParams.get('redirect_path') || '/files';

    try {
      // To test if connection is correct, call RClone to fetch remotes
      setIsLoading(true);
      await new RCloneClient({ endpoint, username, password }).fetchRemotes();

      setIsLoading(false);
      setRCloneInfo({ endpoint, username, password });
      navigate(redirectPath);
    } catch (error) {
      setIsLoading(false);
      setError(error);
    }
  };

  return (
    <LoginHelpDialogProvider>
      <Box className="login-page">
        <Header />
        <Form error={error} isLoading={isLoading} onFormSubmit={handleFormSubmit} />
      </Box>
    </LoginHelpDialogProvider>
  );
}
