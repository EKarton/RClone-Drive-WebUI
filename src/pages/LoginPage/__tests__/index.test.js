import RCloneClient from 'utils/RCloneClient';
import { customRender, userEvent, waitFor, screen } from 'test-utils/react';
import LoginPage from '..';

jest.mock('utils/RCloneClient');

describe('LoginPage', () => {
  const fetchRemotes = jest.fn();
  beforeEach(() => {
    fetchRemotes.mockResolvedValue([]);

    RCloneClient.mockImplementation(() => ({
      fetchRemotes,
    }));
  });

  it('should match snapshot', () => {
    const { baseElement } = customRender(<LoginPage />);

    expect(baseElement).toMatchSnapshot();
  });

  it('should go to the files page when user enters rclone info and clicks Login', async () => {
    const view = customRender(<LoginPage />);

    userEvent.type(screen.getByTestId('endpoint-field'), 'http://localhost:5572');
    userEvent.type(screen.getByTestId('username-field'), 'admin');
    userEvent.type(screen.getByTestId('password-field'), '1234');

    userEvent.click(screen.getByTestId('login-button'));

    await waitFor(() => expect(view.history.location.pathname).toEqual('/files'));
  });

  it('should go to the redirect_path when user enters rclone info and clicks Login', async () => {
    const redirectPath = '/files/T25lZHJpdmU6';
    const route = `/login?redirect_path=${redirectPath}`;
    const view = customRender(<LoginPage />, {}, { route });

    userEvent.type(screen.getByTestId('endpoint-field'), 'http://localhost:5572');
    userEvent.type(screen.getByTestId('username-field'), 'admin');
    userEvent.type(screen.getByTestId('password-field'), '1234');

    userEvent.click(screen.getByTestId('login-button'));

    await waitFor(() => expect(view.history.location.pathname).toEqual(redirectPath));
  });

  it('should show an error message when when user enters rclone info and rclone throws an error', async () => {
    fetchRemotes.mockRejectedValue(new Error('Wrong credentials'));

    customRender(<LoginPage />);

    userEvent.type(screen.getByTestId('endpoint-field'), 'http://localhost:5572');
    userEvent.type(screen.getByTestId('username-field'), 'admin');
    userEvent.type(screen.getByTestId('password-field'), '1234');

    userEvent.click(screen.getByTestId('login-button'));

    await screen.findByText('Wrong credentials');
  });

  it('should open dialog when user clicks on the Help button', async () => {
    customRender(<LoginPage />);

    userEvent.click(screen.getByText('Help'));

    await screen.findByTestId('login-help-dialog');
  });
});
