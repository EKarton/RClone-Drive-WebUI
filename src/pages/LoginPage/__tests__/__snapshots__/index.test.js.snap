// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage should match snapshot 1`] = `
<body>
  <div>
    <div
      class="login-page MuiBox-root css-0"
    >
      <div
        class="login-page-header"
      >
        <span
          class="MuiSwitch-root MuiSwitch-sizeSmall css-dnh7b9-MuiSwitch-root"
        >
          <span
            class="MuiSwitch-switchBase MuiSwitch-colorPrimary MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary PrivateSwitchBase-root css-5ryogn-MuiButtonBase-root-MuiSwitch-switchBase"
          >
            <input
              class="MuiSwitch-input PrivateSwitchBase-input css-1m9pwf3"
              type="checkbox"
            />
            <span
              class="MuiSwitch-thumb css-jsexje-MuiSwitch-thumb"
            />
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </span>
          <span
            class="MuiSwitch-track css-1yjjitx-MuiSwitch-track"
          />
        </span>
      </div>
      <div
        class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root login-page-form css-1v5z4dq-MuiPaper-root-MuiCard-root"
      >
        <form>
          <div
            class="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
          >
            <div
              class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
            >
              <span
                class="MuiTypography-root MuiTypography-h5 MuiCardHeader-title css-1qvr50w-MuiTypography-root"
              >
                Login
              </span>
              <span
                class="MuiTypography-root MuiTypography-body1 MuiCardHeader-subheader css-nrdprl-MuiTypography-root"
              >
                <span
                  class=""
                >
                  Enter your RClone info to get started
                </span>
              </span>
            </div>
          </div>
          <div
            class="MuiCardContent-root login-page-form__contents css-46bh2p-MuiCardContent-root"
          >
            <div
              class="MuiFormControl-root MuiTextField-root login-page-form__text-field css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
              >
                RClone RC Endpoint
                <span
                  aria-hidden="true"
                  class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  data-testid="endpoint-field"
                  name="endpoint"
                  placeholder="Ex: http://localhost:5572"
                  required=""
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-173wfxe"
                  >
                    <span>
                      RClone RC Endpoint
                       
                      *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <div
              class="MuiFormControl-root MuiTextField-root login-page-form__text-field css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
              >
                Username
                <span
                  aria-hidden="true"
                  class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  data-testid="username-field"
                  name="username"
                  required=""
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-173wfxe"
                  >
                    <span>
                      Username
                       
                      *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <div
              class="MuiFormControl-root MuiTextField-root login-page-form__text-field css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
              >
                Password
                <span
                  aria-hidden="true"
                  class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  data-testid="password-field"
                  name="password"
                  required=""
                  type="password"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-173wfxe"
                  >
                    <span>
                      Password
                       
                      *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
          >
            <button
              class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
              data-testid="login-button"
              tabindex="0"
              type="submit"
            >
              Login
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1e6y48t-MuiButtonBase-root-MuiButton-root"
              data-testid="help-button"
              tabindex="0"
              type="button"
            >
              Help
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</body>
`;
